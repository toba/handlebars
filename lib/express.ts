import * as fs from 'fs';
//import * as path from 'path';
import * as Handlebars from 'handlebars';
import { Cache, merge } from '@toba/tools';

export interface ExpressHandlebarsOptions {
   /** Default layout template should be rendered within. */
   defaultLayout: string;
   /** Folder within Express views containing partials. */
   partialsFolder: string;
   /** Folder within Express views containing layouts. */
   layoutsFolder: string;
   /** Whether to cache templates. */
   cacheTemplates: boolean;
}

type RenderCallback = (err: Error, output?: string) => void;

/**
 * Values set in the Express application with the `app.set(name, value)`
 * syntax.
 */
interface ExpressSettings {
   /** Absolute path to renderable views including partials and layouts. */
   views: string;
   filename?: string;
   etag: string;
   /** `NODE_ENV` value if set. */
   env: string;
   'view engine': string;
   'x-powered-by': boolean;
   'trust proxy': boolean;
}

interface RenderContext {
   [key: string]: any;
   /** Cache flag injected by Express. */
   cache?: boolean;
   settings?: ExpressSettings;
   template?: string;
}

const defaultOptions: ExpressHandlebarsOptions = {
   defaultLayout: 'main.hbs',
   partialsFolder: 'partials',
   layoutsFolder: 'layouts',
   cacheTemplates: true
};

/**
 *
 */
export class ExpressHandlebars {
   options: ExpressHandlebarsOptions;
   fileExtension: string;
   cache: Cache<Handlebars.TemplateDelegate<any>>;
   hbs: typeof Handlebars;
   /**
    * @see http://handlebarsjs.com/execution.html
    * @see https://github.com/ericf/express-handlebars/blob/master/lib/express-handlebars.js#L211
    */
   defaultRenderOptions: Handlebars.RuntimeOptions;

   constructor(options: Partial<ExpressHandlebarsOptions> = {}) {
      this.options = merge(defaultOptions, options);
      this.hbs = Handlebars.create();
      this.fileExtension = 'hbs';
      this.cache = new Cache();
      this.renderer = this.renderer.bind(this);
      this.registerHelper = this.registerHelper.bind(this);
   }

   /**
    * Express standard renderer.
    *
    * @example
    *    import { ExpressHandlebars } from '@toba/handlebars';
    *    const ehb = new ExpressHandlebars();
    *    app.engine(ehb.name, ehb.renderer);
    *    app.set('views', './views');
    *    app.set('view engine', ehb.name);
    *
    * @see https://expressjs.com/en/advanced/developing-template-engines.html
    */
   renderer(viewPath: string, context: RenderContext, cb?: RenderCallback) {
      const layout =
         context.template === undefined
            ? this.options.defaultLayout
            : context.template;

      if (layout !== null) {
         // render view within the layout
         return this.renderWithinLayout(layout, viewPath, context, cb);
      }

      if (this.cache.contains(viewPath)) {
         const template = this.cache.get(viewPath);
         cb(null, template(context));
      } else {
         fs.readFile(viewPath, (err, content) => {
            if (err) {
               return cb(err);
            }
            const template = this.hbs.compile(content);
            this.cache.add(viewPath, template);
            cb(null, template(context));
         });
      }
   }

   renderWithinLayout(
      layout: string,
      viewPath: string,
      context: RenderContext,
      cb?: RenderCallback
   ) {
      if (this.cache.contains(viewPath)) {
         const template = this.cache.get(viewPath);
         cb(null, template(context));
      } else {
         fs.readFile(viewPath, (err, content) => {
            if (err) {
               return cb(err);
            }
            const template = this.hbs.compile(content);
            this.cache.add(viewPath, template);
            cb(null, template(context));
         });
      }
   }

   /**
    * Expose useful methods.
    */
   registerHelper(name: string, fn: Handlebars.HelperDelegate) {
      this.hbs.registerHelper(name, fn);
   }
}
