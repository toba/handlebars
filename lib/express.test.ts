import '@toba/test';
import * as path from 'path';
import * as Express from 'express';
import * as request from 'supertest';
import { ExpressHandlebars } from './';
import { HttpStatus } from '@toba/tools';

const viewPath = path.join(__dirname, '__mocks__', 'views');
const app = Express();

beforeAll(() => {
   const ehb = new ExpressHandlebars(viewPath);
   app.set('views', viewPath);
   app.set('view engine', ehb.fileExtension);
   app.engine(ehb.fileExtension, ehb.renderer);
});

function makeRoute(path: string, viewName: string, layout?: string) {
   app.get(path, (_req: Express.Request, res: Express.Response) => {
      res.render(
         viewName,
         { key: 'value', layout },
         (err: Error, html: string) => {
            expect(err).toBeNull();
            res.write(html);
            res.end();
         }
      );
   });
}

test(
   'renders without layout',
   async () => {
      makeRoute('/', 'home', null);
      const res = await request(app).get('/');
      expect(res.status).toBe(HttpStatus.OK);
      expect(res.text).toMatchSnapshot();
   },
   10000
);

test(
   'renders within layout',
   async () => {
      makeRoute('/home', 'home');
      const res = await request(app).get('/home');
      expect(res.status).toBe(HttpStatus.OK);
      expect(res.text).toMatchSnapshot();
   },
   10000
);
