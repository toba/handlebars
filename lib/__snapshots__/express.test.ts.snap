// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders data 1`] = `
"<h1>Mockery</h1>
<ol>
   <li>one</li>
   <li>two</li>
   <li>three</li>
</ol>"
`;

exports[`renders data within layout 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
   <meta value=\\"something\\" /> 
</head>

<body>
   <nav>
   Navigation
</nav> <h1>Mockery</h1>
<ol>
   <li>one</li>
   <li>two</li>
   <li>three</li>
</ol> <footer>
   Footer
</footer>
</body>

</html>"
`;

exports[`renders within layout 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
   <meta value=\\"something\\" />    <meta name=\\"content-for\\" />

   <meta name=\\"content-for\\" />

</head>

<body>
   <nav>
   Navigation
</nav> 
<article>
   <h1>Home</h1>
</article> <footer>
   Footer
</footer>
</body>

</html>"
`;

exports[`renders without layout 1`] = `
"
<article>
   <h1>Home</h1>
</article>"
`;
